proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8372,143 +8394,26 @@ lemma MIN_BUDGET_lower_bounds[elim!]:
-   apply (case_tac "sc_period y = 0"; clarsimp)
-   apply (rename_tac s n sc h t)
-
-   apply (prop_tac "unat (r_amount h) + unat (r_amount t) \<le> unat max_time")
-    apply (rule_tac y="unat (r_time h) + unat (r_amount h) + unat (r_amount t)" in order_trans, simp)
-    apply (rule_tac y="unat (r_time t) + unat (r_amount t)" in order_trans, simp)
-    apply (clarsimp simp: no_overflow_def)
-   apply (prop_tac "unat (r_amount h) \<le> unat MAX_SC_PERIOD")
-    apply (rule_tac y="unat (r_amount h + r_amount t)" in order_trans)
-     apply (subst unat_add_lem', overflow_hammer)
-     apply simp
-    apply (simp, overflow_hammer)
-   apply (prop_tac "unat (r_amount t) \<le> unat MAX_SC_PERIOD")
-    apply (rule_tac y="unat (r_amount h + r_amount t)" in order_trans)
-     apply (subst unat_add_lem', overflow_hammer)
-     apply simp
-    apply (simp, overflow_hammer)
-   apply (prop_tac "usage \<le> r_amount h")
-    apply (erule_tac order_trans[rotated], overflow_hammer)
-    apply (erule add_left_mono_trans)
-    apply (erule_tac order_trans[rotated])
-    apply (rule iffD1[OF word_le_nat_alt], fastforce)
-   apply (prop_tac "unat (r_amount h - usage) \<le> unat MAX_SC_PERIOD")
-    apply (erule_tac y="unat (r_amount h)" in order_trans[rotated])
-    apply (rule iffD1[OF word_le_nat_alt], overflow_hammer)
-   apply (prop_tac "unat MIN_BUDGET \<le> unat MAX_SC_PERIOD")
-    apply (rule_tac y="unat (sc_budget sc)" in order_trans[rotated])
-     apply (rule iffD1[OF word_le_nat_alt], overflow_hammer)
-    apply (rule iffD1[OF word_le_nat_alt], fastforce)
-
-   apply (intro conjI allI impI)
-     apply (rule unat_le_mono)
-     apply (rule word_le_minus_mono_left)
-      apply (simp add: word_le_nat_alt)
-     apply simp
-    \<comment> \<open>no overflow\<close>
-      apply (clarsimp simp: no_overflow_def)
-      apply (subst unat_split_plus[symmetric], simp)
-       apply (meson unat_sum_bound_equiv order_trans_rules(23) word_le_nat_alt word_le_imp_diff_le)
-      apply simp
-     apply (rule word_sub_move)
-      apply (simp add: Groups.add_ac(2))
-     using unat_sum_bound_equiv apply fastforce
-    apply (metis (no_types, hide_lams) Groups.add_ac(2) add_diff_cancel_right' unat_sum_bound_equiv le_no_overflow olen_add_eqv plus_minus_no_overflow_ab)
-      \<comment> \<open>window\<close>
-   apply (clarsimp simp: window_def)
-   apply (subst unat_split_plus[symmetric])
-    apply clarsimp
-    apply (rule_tac y="r_time t" in order_trans)
-     apply (meson order_trans_rules(23) word_le_nat_alt word_sub_le)
-    using unat_sum_bound_equiv no_overflow_length2 apply blast
-   apply clarsimp
-   using unat_plus_gt_trans apply blast
+  apply (clarsimp simp: refill_budget_check_round_robin_def)
   apply (wpsimp wp: set_refills_wp)
-  apply (clarsimp simp: vs_all_heap_simps valid_refills_def rr_valid_refills_def3 obj_at_def MIN_REFILLS_def
-                        ordered_disjoint_def no_overflow_def window_def
+  apply (case_tac "p = cur_sc s"; simp?)
```

Comment:
Maybe just go `case_tac "p ~= cur_sc s"`, rather than the case_tac then prefer 2?